core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java

```diff
@@ -50,12 +61,61 @@ public static GrpcManagedChannelPool INSTANCE() {
   /** Used to control access to mChannel */
   private ReentrantReadWriteLock mLock;
 
+  private HashSet<ChannelKey> mChannelsToRelease;
+  /** Scheduler for destruction of released channels. */
+  protected final ScheduledExecutorService mScheduler;
+
   /**
    * Creates a new {@link GrpcManagedChannelPool}.
    */
   public GrpcManagedChannelPool() {
     mChannels = new HashMap<>();
     mLock = new ReentrantReadWriteLock(true);
+
+    mChannelsToRelease = new HashSet<>();
+    mScheduler = Executors.newScheduledThreadPool(1,
+            ThreadFactoryUtils.build("grpc-channel-terminator", true));
+
+    // Channel termination callback will be fired in the same interval as
+    // the channel shutdown timeout.
+    long channelTerminationIntervalMs =
+        Configuration.getMs(PropertyKey.MASTER_GRPC_CHANNEL_SHUTDOWN_TIMEOUT);
+    mScheduler.scheduleAtFixedRate(() -> {
+      destroyChannels();
+    }, channelTerminationIntervalMs, channelTerminationIntervalMs, TimeUnit.MILLISECONDS);
+  }
+
+  private void destroyChannels() {
+    try (LockResource lockExclusive = new LockResource(mLock.writeLock())) {
+      int iDestroyed = 0, iReactivated = 0;
+      for (ChannelKey channelKey : mChannelsToRelease) {
+        Verify.verify(mChannels.containsKey(channelKey));
+        ManagedChannelReference channelRef = mChannels.get(channelKey);
+        channelRef.reference();
+        if (!channelRef.dereference()) {
+          iReactivated++;
+        } else {
+          iDestroyed++;
+          ManagedChannel channel = channelRef.reference();
+          channel.shutdown();
+          try {
+            channel.awaitTermination(
+                Configuration.getMs(PropertyKey.MASTER_GRPC_CHANNEL_SHUTDOWN_TIMEOUT),
+                TimeUnit.MILLISECONDS);
```

Comment:
It seems inefficient to wait for channel termination while holding the resource lock? Ideally we can just remove the channel from `mChannels` and release the lock.

Reply:
I don't think this comment is addressed yet. Is there any way someone can get a reference to the channel after it has been removed from `mChannels`? If not, I think we can just remove from `mChannels` while holding the lock and then destroy the channels afterward?

Reply:
It is already deleted from the mChannels after successful destruction.
However channels that failed to destruct should not be served again so moving the deletion to before attempting the destruction.

Is this good enough?

Reply:
My point is we are holding the exclusive lock while we wait for the channels to be destroyed. That seems unnecessary since we logically destroy the channel by removing it from the map (no one else can see the channel anymore), which is cheaper than waiting for termination. Once we have logically removed all the channels we want to destroy later, we can relinquish the lock since we have exclusive access to the channels we want to destroy.

Reply:
Makes sense. Will do.